### PR TITLE
Change etcd Get operation to serializable

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref https://github.com/kubernetes/kubernetes/issues/34709

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34711)

### CODE REVIEW

**Comment by wojtek-t** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by wojtek-t** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by xiang90** [comment]
This makes a lot of sense for 1 node cluster, but not multiple nodes (we might see time go backward if API server contact a different etcd node due to previous etcd node failure). For multiple nodes, we probably want quorum read by default. If we specify a revision, then s-read is perfectly fine since we know what we want to get. Also there is an on going effort to make quorum read MUCH cheaper in etcd 3.1. I am not super worry about this.

**Comment by wojtek-t** [comment]
@xiang90 - what do you think about flag-gating this option?

**Comment by wojtek-t** [comment]
Obviously, in this case we should set default to false. But that would probably make it possible to test it (and see if this is a problem).

**Comment by xiang90** [comment]
@wojtek-t Yea. If it makes testing easier, add a flag and disable it by default.

**Comment by wojtek-t** [comment]
Thanks - will update the PR tomorrow.

**Comment by wojtek-t** [comment]
@xiang90 - PTAL

I decided to you the existing "Quorum" parameter to configure it - I think consistency here is pretty important. WDYT?

**Comment by wojtek-t** [comment]
@hongchaodeng - if you will be online, maybe you can take a look?

File: pkg/storage/etcd3/store.go [hunk 1]
```diff
@@ -58,19 +61,25 @@ type objState struct {
 }
 
 // New returns an etcd3 implementation of storage.Interface.
-func New(c *clientv3.Client, codec runtime.Codec, prefix string) storage.Interface {
-	return newStore(c, codec, prefix)
+func New(c *clientv3.Client, quorumRead bool, codec runtime.Codec, prefix string) storage.Interface {
```
Comment (xiang90) [inline pkg/storage/etcd3/store.go]: This flag is mainly used for testing (or special environment when user knows they are running 1 member etcd cluster), and it requires a recompile. I do not really like to use bool as arg and want to keep New as simple as possible. So can we not modify the New func to keep it clean, and add another NewWithNoQuorumRead func?

Comment (wojtek-t) [inline pkg/storage/etcd3/store.go]: @xiang90 - I added the "NewWithNoQuorumRead" func - I think it makes perfect sense.

However, I completely don't understand your argument about "recompile" - this is being set using this flag:
https://github.com/kubernetes/kubernetes/blob/master/pkg/genericapiserver/options/etcd_options.go#L49
and doesn't require recompile.
I think we shouldn't introduce another flag for the same thing (I mean quorum read from etcd).

**Comment by wojtek-t** [comment]
@xiang90 - thanks a lot for taking a look into it during weekend!

**Comment by xiang90** [comment]
@wojtek-t I was not aware of that we already have a external flag there... I will probably add some warning for the flag... It seems to be dangerous when people running multiple nodes etcd cluster + multiple API servers.

**Comment by wojtek-t** [comment]
@xiang90 - I'm happy to add a warning in a separate PR. However, I would like to clarify what do you mean be that - a log? a comment? documentation? or something else?

**Pull request merged**